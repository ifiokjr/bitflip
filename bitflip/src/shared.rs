pub struct AppError(anyhow::Error);

// Tell axum how to convert `AppError` into a response.
#[cfg(feature = "ssr")]
impl axum::response::IntoResponse for AppError {
	fn into_response(self) -> axum::response::Response {
		(
			http::StatusCode::INTERNAL_SERVER_ERROR,
			format!("App error: {}", self.0),
		)
			.into_response()
	}
}

// This enables using `?` on functions that return `Result<_, anyhow::Error>` to
// turn them into `Result<_, AppError>`. That way you don't need to do that
// manually.
impl<E> From<E> for AppError
where
	E: Into<anyhow::Error>,
{
	fn from(err: E) -> Self {
		Self(err.into())
	}
}
