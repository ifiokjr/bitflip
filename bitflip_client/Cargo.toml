[package]
name = "bitflip_client"
version = { workspace = true }
authors = { workspace = true }
edition = { workspace = true }
homepage = { workspace = true }
license = { workspace = true }
repository = { workspace = true }
rust-version = { workspace = true }
description = "The rust based client for interacting with the bitflip solana program `bitflip`."

[lints]
workspace = true

[lib]
crate-type = ["cdylib", "lib"]
name = "bitflip_client"

[dependencies]
anchor-lang = { workspace = true }
anchor-spl = { workspace = true }
async-trait = { workspace = true }
bitflip_program = { workspace = true, features = ["client"] }
derive_more = { workspace = true, features = ["debug", "deref", "deref_mut"] }
futures = { workspace = true }
log = { workspace = true }
paste = { workspace = true }
solana-program = { workspace = true }
solana-sdk = { workspace = true }
typed-builder = { workspace = true }
wallet_standard = { workspace = true, features = ["solana"] }
wasm_client_anchor = { workspace = true }
wasm_client_solana = { workspace = true }

[dev-dependencies]
anyhow = { workspace = true }
assert2 = { workspace = true }
insta = { workspace = true, features = ["ron", "yaml", "redactions"] }
rstest = { workspace = true }
solana-program-test = { workspace = true }
spl-associated-token-account = { workspace = true }
spl-transfer-hook-interface = { workspace = true }
test-log = { workspace = true, features = ["log", "trace"] }
test_utils = { workspace = true }
test_utils_solana = { workspace = true, features = ["test_validator"] }
tokio = { workspace = true, features = ["test-util", "rt", "macros", "time"] }
wallet_standard_wallets = { workspace = true }

[features]
ssr = ["wasm_client_solana/ssr"]
js = [
	"wasm_client_solana/js",
	"wasm_client_anchor/js",
	"bitflip_program/no-entrypoint",
]
