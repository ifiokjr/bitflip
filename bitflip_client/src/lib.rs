pub use crate::client::*;
pub use crate::pda::*;

pub mod client;
mod pda;
pub mod prelude {
	pub use wasm_client_anchor::prelude::*;
}
