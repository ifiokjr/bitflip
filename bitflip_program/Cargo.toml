[package]
name = "bitflip_program"
version = "0.0.0"
edition = "2021"
description = "Created with Anchor"

[lints]
workspace = true

[lib]
crate-type = ["cdylib", "lib"]
name = "bitflip_program"

[features]
default = ["cpi"]
cpi = []
no-entrypoint = []
no-idl = []
no-log-ix-name = []
idl-build = ["anchor-lang/idl-build", "anchor-spl/idl-build"]
serde = ["dep:serde", "dep:serde_with"]
client = ["serde", "no-entrypoint", "dep:typed-builder"]

[dependencies]
anchor-lang = { workspace = true, features = ["interface-instructions", "init-if-needed"] }
anchor-spl = { workspace = true }
bitflags = { workspace = true }
bytemuck = { workspace = true }
const_str_to_pubkey = { workspace = true }
derive_more = { workspace = true, features = ["add", "add_assign"] }
serde = { workspace = true, features = ["derive"], optional = true }
serde-big-array = { workspace = true }
serde_with = { workspace = true, optional = true }
typed-builder = { workspace = true, optional = true }

[dev-dependencies]
anyhow = { workspace = true }
assert2 = { workspace = true }
insta = { workspace = true, features = ["ron", "yaml", "redactions"] }
rstest = { workspace = true }
test_utils = { workspace = true }
test_utils_solana = { workspace = true }
tokio = { workspace = true, features = ["test-util", "rt", "macros", "time"] }
