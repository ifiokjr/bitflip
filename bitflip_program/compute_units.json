{
	"instructions": {
		"config_initialize": {
			"compute_units": 17569,
			"rounded_compute_units": 20000,
			"description": "Initialize the config"
		},
		"config_update_authority": {
			"compute_units": 2326,
			"rounded_compute_units": 10000,
			"description": "Update the authority of the config"
		},
		"flip_bit_test": {
			"compute_units": 64832,
			"rounded_compute_units": 70000,
			"description": "Flip a single bit from `0` to `1`"
		},
		"game_initialize": {
			"compute_units": 12979,
			"rounded_compute_units": 20000,
			"description": "Initialize the game"
		},
		"game_refresh_signer": {
			"compute_units": 2232,
			"rounded_compute_units": 10000,
			"description": "Refresh the signer"
		},
		"token_group_initialize": {
			"compute_units": 47429,
			"rounded_compute_units": 50000,
			"description": "Initialize the token group"
		},
		"token_initialize:bit": {
			"compute_units": 63184,
			"rounded_compute_units": 70000,
			"description": "Initialize group token: Bit"
		},
		"token_initialize:gibibit": {
			"compute_units": 59746,
			"rounded_compute_units": 70000,
			"description": "Initialize member token: GibiBit"
		},
		"token_initialize:kibibit": {
			"compute_units": 59745,
			"rounded_compute_units": 70000,
			"description": "Initialize member token: KibiBit"
		},
		"token_initialize:mebibit": {
			"compute_units": 59742,
			"rounded_compute_units": 70000,
			"description": "Initialize member token: MebiBit"
		}
	},
	"version": "0.0.0"
}
