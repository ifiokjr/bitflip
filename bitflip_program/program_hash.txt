ebc22d77bc8ddd23f326598673e9fe7aa99fcdce9572fb8559a8d9782ade4ab8
