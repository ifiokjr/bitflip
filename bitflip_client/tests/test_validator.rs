#![cfg(feature = "test_validator")]
use std::collections::HashMap;

use anchor_spl::token_2022;
use assert2::check;
use bitflip_client::BitflipProgramClient;
use bitflip_client::get_pda_bits_data_section;
use bitflip_client::get_pda_bits_meta;
use bitflip_client::get_pda_config;
use bitflip_client::get_pda_mint;
use bitflip_client::get_pda_treasury;
use bitflip_client::initialize_bits_data_sections_request;
use bitflip_client::initialize_token_request;
use bitflip_program::BITS_DATA_SECTION_LENGTH;
use bitflip_program::BitsDataSectionState;
use bitflip_program::BitsMetaState;
use bitflip_program::ConfigState;
use bitflip_program::ID_CONST;
use bitflip_program::InitializeConfigProps;
use bitflip_program::TOKEN_DECIMALS;
use bitflip_program::accounts::InitializeConfig;
use futures::future::try_join_all;
use shared::TestBitflipProgramClient;
use shared::create_admin_keypair;
use shared::create_authority_keypair;
use shared::create_bits_meta_state;
use shared::create_config_state;
use shared::create_runner;
use shared::create_runner_with_accounts;
use shared::get_admin_program;
use solana_sdk::signer::Signer;
use solana_sdk::system_program;
use spl_associated_token_account::get_associated_token_address_with_program_id;
use test_log::test;
use test_utils::create_insta_redaction;
use test_utils_solana::prelude::*;
use wallet_standard_wallets::MemoryWallet;
use wasm_client_solana::solana_account_decoder::parse_account_data::SplTokenAdditionalData;
use wasm_client_solana::solana_account_decoder::parse_token::parse_token_v2;

mod shared;

#[test(tokio::test)]
async fn initialize_config_test() -> anyhow::Result<()> {
	let admin_keypair = create_admin_keypair();
	let admin = admin_keypair.pubkey();
	let authority_keypair = create_authority_keypair();
	let authority = authority_keypair.pubkey();
	let system_program = system_program::id();
	let (config, _) = get_pda_config();
	let (treasury, _) = get_pda_treasury();
	let runner = create_runner().await;
	let rpc = runner.rpc();
	let wallet = MemoryWallet::new(rpc.clone(), &[admin_keypair.insecure_clone()]);
	let bitflip_program_client: TestBitflipProgramClient = BitflipProgramClient::builder()
		.rpc(rpc.clone())
		.wallet(wallet)
		.build()
		.into();

	let initialize_config_request = bitflip_program_client
		.initialize_config()
		.args(InitializeConfigProps { authority })
		.accounts(InitializeConfig {
			config,
			admin,
			treasury,
			system_program,
		})
		.signers(vec![&admin_keypair])
		.build();

	let simulation = initialize_config_request.simulate_transaction().await?;

	log::info!("units_consumed: {simulation:#?}");
	check!(simulation.value.units_consumed.unwrap() < 75_000);

	initialize_config_request
		.sign_and_send_transaction()
		.await?;

	let config_state_account = bitflip_program_client
		.account::<ConfigState>(&config)
		.await?;

	insta::assert_yaml_snapshot!(config_state_account);

	Ok(())
}

#[test(tokio::test)]
async fn initialize_token() -> anyhow::Result<()> {
	let mut accounts = HashMap::new();
	let admin_keypair = create_admin_keypair();
	let (config, _) = get_pda_config();
	let authority_keypair = create_authority_keypair();
	let (mint, _) = get_pda_mint();
	let (treasury, _) = get_pda_treasury();
	let token_program = token_2022::ID;
	let treasury_token_account =
		get_associated_token_address_with_program_id(&treasury, &mint, &token_program);

	accounts.insert(config, create_config_state(None));

	let runner = create_runner_with_accounts(accounts).await;
	let rpc = runner.rpc();

	let wallet = MemoryWallet::new(rpc.clone(), &[admin_keypair.insecure_clone()]);
	let bitflip_program_client: TestBitflipProgramClient = BitflipProgramClient::builder()
		.rpc(rpc.clone())
		.wallet(wallet)
		.build()
		.into();

	let request = initialize_token_request(&bitflip_program_client, &authority_keypair);

	let simulation = request.simulate_transaction().await?;

	log::info!("units_consumed: {simulation:#?}");
	check!(simulation.value.units_consumed.unwrap() < 200_000);

	let signature = request.sign_and_send_transaction().await?;
	rpc.confirm_transaction(&signature).await?;

	let authority_redaction = create_insta_redaction(treasury, "authority:pubkey");
	let mint_redaction = create_insta_redaction(mint, "mint:pubkey");
	let program_redaction = create_insta_redaction(ID_CONST, "program:pubkey");
	let mint_data = rpc.get_account_data(&mint).await?;
	let mint_account = parse_token_v2(
		&mint_data,
		Some(&SplTokenAdditionalData::with_decimals(TOKEN_DECIMALS)),
	)?;

	log::info!("mint account: {mint_account:#?}");

	insta::assert_yaml_snapshot!(mint_account, {
		".info.mintAuthority" => insta::dynamic_redaction(authority_redaction.clone()),
		".info.freezeAuthority" => insta::dynamic_redaction(authority_redaction.clone()),
		".info.extensions[0].state.authority" => insta::dynamic_redaction(authority_redaction.clone()),
		".info.extensions[0].state.metadataAddress" =>	insta::dynamic_redaction(mint_redaction.clone()),
	  ".info.extensions[1].state.closeAuthority" => insta::dynamic_redaction(authority_redaction.clone()),
		".info.extensions[2].state.authority" => insta::dynamic_redaction(authority_redaction.clone()),
	  ".info.extensions[2].state.programId" => insta::dynamic_redaction(program_redaction),
		".info.extensions[3].state.delegate" =>	insta::dynamic_redaction(authority_redaction.clone()),
	  ".info.extensions[4].state.updateAuthority" => insta::dynamic_redaction(authority_redaction.clone()),
	  ".info.extensions[4].state.mint" =>	insta::dynamic_redaction(mint_redaction.clone())
	});

	let treasury_token_account_data = rpc.get_account_data(&treasury_token_account).await?;
	let token_account = parse_token_v2(
		&treasury_token_account_data,
		Some(&SplTokenAdditionalData::with_decimals(TOKEN_DECIMALS)),
	)?;

	log::info!("token account: {token_account:#?}");
	insta::assert_yaml_snapshot!(token_account, {
		".info.mint" => insta::dynamic_redaction(mint_redaction),
		".info.owner" => insta::dynamic_redaction(authority_redaction),
	});

	Ok(())
}

#[test(tokio::test)]
async fn initialize_bits_meta() -> anyhow::Result<()> {
	let mut accounts = HashMap::new();
	let (config, _) = get_pda_config();
	let authority_keypair = create_authority_keypair();
	let authority = authority_keypair.pubkey();
	let (bits_meta, _) = get_pda_bits_meta(0);
	let system_program = system_program::ID;

	accounts.insert(config, create_config_state(None));

	let runner = create_runner_with_accounts(accounts).await;
	let rpc = runner.rpc();
	let bitflip_program_client = get_admin_program(rpc);

	let initialize_bits_meta_request = bitflip_program_client
		.initialize_bits_meta()
		.accounts(bitflip_program::accounts::InitializeBitsMeta {
			config,
			bits_meta,
			authority,
			system_program,
		})
		.signers(vec![&authority_keypair])
		.build();

	let simulation = initialize_bits_meta_request.simulate_transaction().await?;
	log::info!("simulation: {simulation:#?}");

	initialize_bits_meta_request
		.sign_and_send_transaction()
		.await?;

	let bits_meta_account: BitsMetaState = bitflip_program_client.account(&bits_meta).await?;

	insta::assert_yaml_snapshot!(bits_meta_account);

	Ok(())
}

#[test(tokio::test)]
async fn initialize_bits_data_sections() -> anyhow::Result<()> {
	let index = 0;
	let (config, _) = get_pda_config();
	let mut accounts = HashMap::new();
	let authority_keypair = create_authority_keypair();
	let (bits_meta, _) = get_pda_bits_meta(index);
	accounts.insert(config, create_config_state(None));
	accounts.insert(bits_meta, create_bits_meta_state(index));

	let runner = create_runner_with_accounts(accounts).await;
	let rpc = runner.rpc();
	let program_client = get_admin_program(rpc);
	let program_client_ref = &program_client;

	let requests =
		initialize_bits_data_sections_request(&program_client, &authority_keypair, index).await?;

	for request in &requests {
		request.sign_and_send_transaction().await?;
	}

	let range = 0u8..16u8;
	let futures = range.into_iter().map(|section| {
		async move {
			let pubkey = get_pda_bits_data_section(index, section).0;
			program_client_ref
				.account::<BitsDataSectionState>(&pubkey)
				.await
		}
	});
	let accounts = try_join_all(futures).await?;

	for (section, account) in accounts.iter().enumerate() {
		let bump = get_pda_bits_data_section(index, section as u8).1;
		check!(
			account.data.len() == BITS_DATA_SECTION_LENGTH,
			"check data length for section:  {section}"
		);
		check!(
			account.section == section as u8,
			"check section:  {section}"
		);
		check!(account.bump == bump, "check bump for section:  {section}");
	}

	Ok(())
}
