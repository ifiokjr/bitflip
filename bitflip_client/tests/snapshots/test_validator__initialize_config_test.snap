---
source: bitflip_client/tests/test_validator.rs
expression: config_state_account
---
authority: EjxoynXa5jTq4MyZzWtTz6Uq8UHQDxebkWhvEkDXoZxH
lamportsPerBit: 100000
bump: 254
treasuryBump: 255
mintBump: 0
bitsIndex: 0
